File: mac/Classes/Model/Actions/Specific/CompileFileAction.m

#import "CompileFileAction.h"
#import "ScriptInvocationStep.h"
#import "Project.h"
//#import "Plugin.h"
#import "LRProjectFile.h"
#import "LRPathProcessing.h"
#import "LRFileTargetResult.h"


@implementation CompileFileAction

- (NSString *)label {
    return self.type.name;
}

- (void)loadFromMemento:(NSDictionary *)memento {
    [super loadFromMemento:memento];
    self.compilerName = memento[@"compiler"];
//    self.intrinsicInputPathSpec = [ATPathSpec pathSpecWithString:@"*.css" syntaxOptions:ATPathSpecSyntaxFlavorExtended];
    _outputFilterOption = [FilterOption filterOptionWithMemento:(memento[@"output"] ?: @"subdir:.")];

    NSString *inputFilter = self.type.manifest[@"input"];
    self.intrinsicInputPathSpec = [ATPathSpec pathSpecWithString:inputFilter syntaxOptions:ATPathSpecSyntaxFlavorExtended];
}

- (void)setOutputFilterOption:(FilterOption *)outputFilterOption {
    if (_outputFilterOption != outputFilterOption) {
        _outputFilterOption = outputFilterOption;
        [self didChange];
    }
}

- (void)updateMemento:(NSMutableDictionary *)memento {
    [super updateMemento:memento];

    memento[@"output"] = self.outputFilterOption.memento;
}

- (LRProjectFile *)destinationFileForSourceFile:(LRProjectFile *)file inProject:(Project *)project {
    NSString *destinationName = LRDeriveDestinationFileName([file.relativePath lastPathComponent], self.type.manifest[@"output"], self.intrinsicInputPathSpec);

    BOOL outputMappingIsRecursive = YES; // TODO: make this conditional
    if (outputMappingIsRecursive) {
        NSUInteger folderComponentCount = self.inputFilterOption.folderComponentCount;
        if (folderComponentCount > 0) {
            NSArray *components = [[file.relativePath stringByDeletingLastPathComponent] pathComponents];
            if (components.count > folderComponentCount) {
                destinationName = [[[components subarrayWithRange:NSMakeRange(folderComponentCount, components.count - folderComponentCount)] componentsJoinedByString:@"/"] stringByAppendingPathComponent:destinationName];
            }
        }
    }

    NSString *destinationRelativePath = nil;
    if (self.outputFilterOption.subfolder)
        destinationRelativePath = [self.outputFilterOption.subfolder stringByAppendingPathComponent:destinationName];

    return [LRProjectFile fileWithRelativePath:destinationRelativePath project:project];
}

- (void)handleDeletionOfFile:(LRProjectFile *)file inProject:(Project *)project {
    LRProjectFile *destinationFile = [self destinationFileForSourceFile:file inProject:project];
    if (![destinationFile.absoluteURL isEqual:file.absoluteURL] && destinationFile.exists) {
        [[NSFileManager defaultManager] removeItemAtURL:destinationFile.absoluteURL error:NULL];
    }
}

- (void)configureStep:(ScriptInvocationStep *)step forFile:(LRProjectFile *)file {
    [super configureStep:step forFile:file];

    [step addFileValue:file forSubstitutionKey:@"src"];

    LRProjectFile *destinationFile = [self destinationFileForSourceFile:file inProject:step.project];

    NSURL *destinationFolderURL = [destinationFile.absoluteURL URLByDeletingLastPathComponent];
    if (![destinationFolderURL checkResourceIsReachableAndReturnError:NULL]) {
        [[NSFileManager defaultManager] createDirectoryAtURL:destinationFolderURL withIntermediateDirectories:YES attributes:nil error:NULL];
    }
    [step addFileValue:destinationFile forSubstitutionKey:@"dst"];
}

- (void)didCompleteCompilationStep:(ScriptInvocationStep *)step forFile:(LRProjectFile *)file {
    LRProjectFile *outputFile = [step fileForKey:@"dst"];
    [file.project hackhack_didWriteCompiledFile:outputFile];
}

- (BOOL)supportsFileTargets {
    return YES;
}

- (LRTargetResult *)fileTargetForRootFile:(LRProjectFile *)sourceFile {
    return [[LRFileTargetResult alloc] initWithAction:self sourceFile:sourceFile];
}

@end

File: mac/Classes/Model/Actions/Specific/CompileFileAction.swift

import Foundation

@objc class CompileFileAction : ScriptInvocationAction {

    var compilerName : String?

    var outputFilterOption : FilterOption = FilterOption(memento: "subdir.") {
        didSet {
            if outputFilterOption != oldValue {
                didChange();
            }
        }
    }

    override var label : String {
        return type.name
    }

    override func loadFromMemento(memento: NSDictionary!) {
        super.loadFromMemento(memento)
        compilerName = memento["compiler"] as? String
        outputFilterOption = FilterOption(memento: NV(memento["output"] as? String, "subdir:."))
        intrinsicInputPathSpec = type.combinedIntrinsicInputPathSpec
    }

    override func updateMemento(memento: NSMutableDictionary!) {
        super.updateMemento(memento)
        memento["output"] = outputFilterOption.memento
    }

    func destinationFileForSourceFile(file: LRProjectFile, inProject project: Project) -> LRProjectFile {
        var destinationName = LRDeriveDestinationFileName(file.relativePath.lastPathComponent, type.manifest["output"]! as String, intrinsicInputPathSpec)

        var outputMappingIsRecursive = true  // TODO: make this conditional
        if outputMappingIsRecursive {
            let folderComponentCount = inputFilterOption.folderComponentCount
            if folderComponentCount > 0 {
                let components = file.relativePath.stringByDeletingLastPathComponent.pathComponents
                if components.count > folderComponentCount {
                    destinationName = join("/", components[folderComponentCount .. components.count]).stringByAppendingPathComponent(destinationName)
                }
            }
        }

        let destinationRelativePath = outputFilterOption.subfolder!.stringByAppendingPathComponent(destinationName)
        return LRProjectFile(relativePath: destinationRelativePath, project: project)
    }

    override func handleDeletionOfFile(file: LRProjectFile!, inProject project: Project!) {
        let destinationFile = destinationFileForSourceFile(file, inProject: project)
        if (destinationFile.absoluteURL != file.absoluteURL) && destinationFile.exists {
            NSFileManager.defaultManager().removeItemAtURL(destinationFile.absoluteURL, error:nil)
        }
    }

    override func configureStep(step: ScriptInvocationStep!, forFile file: LRProjectFile!) {
        super.configureStep(step, forFile: file)

        step.addFileValue(file, forSubstitutionKey: "src")

        let destinationFile = destinationFileForSourceFile(file, inProject: step.project)
        step.addFileValue(destinationFile, forSubstitutionKey: "dst")

        let destinationFolderURL = destinationFile.absoluteURL.URLByDeletingLastPathComponent
        if !destinationFolderURL.checkResourceIsReachableAndReturnError(nil) {
            NSFileManager.defaultManager().createDirectoryAtURL(destinationFolderURL, withIntermediateDirectories: true, attributes: nil, error: nil)
        }
    }

    override func didCompleteCompilationStep(step: ScriptInvocationStep!, forFile file: LRProjectFile!) {
        let outputFile = step.fileForKey("dst")
        file.project.hackhack_didWriteCompiledFile(outputFile)
    }

    
    override var supportsFileTargets : Bool {
        return true
    }

    override func fileTargetForRootFile(file: LRProjectFile!) -> LRTarget! {
        return LRFileTarget(action: self, sourceFile: file)
    }

}
